### PR TITLE
Segfaults in TString(char, Ssiz_t) fix

### PR DESCRIPTION
**Pull request by AdvaitDhingra**
I resolved the issue described in [this](https://github.com/root-project/root/issues/6748) issue.

Instead of throwing an error in 

```
root [0] Ssiz_t n = -5;
root [1] TString *a = new TString("a", n);
```
It now creates an empty String by doing this:
```
char *data = Init(0, 0);
memcpy(data, "", 0);
return;
```
Is this correct?

### CODE REVIEW

File: core/base/src/TString.cxx [hunk 1]
```diff
@@ -126,7 +126,8 @@ TString::TString(const std::string &s)
 TString::TString(const char *cs, Ssiz_t n)
 {
    if (n < 0) {
-      Error("TString::TString", "Negative length!");
+      char *data = Init(0, 0);
```
Comment (jblomer) [inline core/base/src/TString.cxx:129]: I think the `Zero()` member method is more concise to initialize the empty string.

**Review by jblomer — CHANGES_REQUESTED** [review]
Many thanks!

In order to address #6748 fully, could you please also take a look at the other constructors that can take negative lengths.

It would also be good to accompany the change by a unit test. You can create a new file core/base/test/TString.cxx (use, e.g., TNamedTests.cxx as a blue print) to exercise the construction of `TString`.  The new unit test would need to be added to the core/base/test/CMakeLists.txt file.

**Comment by linev** [comment]
Please rebase your branch and remove merge commit.

**Comment by AdvaitDhingra** [comment]
> Please rebase your branch and remove merge commit.

How do I do that? Sorry, I'm new to git. 

I did that merge to keep my fork up to date.

**Comment by linev** [comment]
> How do I do that? Sorry, I'm new to git.

Normally you need to get latest ROOT master and do on your branch:
```
git fetch upstream
git rebase upstream/master
git push --force
```
`upstream` is typical name for remote for main ROOT repository.
But it may fail while you add merge commit.
In this case just do cherry-pick of all **your** commits like:
```
git fetch upstream
git reset --hard upstream/master
git cherry-pick 28b6306
git cherry-pick 4d1369a 
...
git push --force
```

**Comment by AdvaitDhingra** [comment]
Ok, thanks. I will do this

**Comment by AdvaitDhingra** [comment]
> > How do I do that? Sorry, I'm new to git.
> 
> Normally you need to get latest ROOT master and do on your branch:
> 
> ```
> git fetch upstream
> git rebase upstream/master
> git push --force
> ```
> 
> `upstream` is typical name for remote for main ROOT repository.
> But it may fail while you add merge commit.
> In this case just do cherry-pick of all **your** commits like:
> 
> ```
> git fetch upstream
> git reset --hard upstream/master
> git cherry-pick 28b6306
So I just do this to all **my** commits, if I understand correctly?
> git cherry-pick 4d1369a

**Comment by AdvaitDhingra** [comment]
Have I done the cherry-picks correctly? Sorry if it's wrong, i'm a bit new 😅

**Comment by AdvaitDhingra** [comment]
Am I doing this correctly? I am a bit confused. I have never used github in the terminal in this way before

File: core/base/src/TString.cxx [hunk 2]
```diff
@@ -238,6 +243,11 @@ TString::~TString()
 
 char *TString::Init(Ssiz_t capacity, Ssiz_t nchar)
 {
+	if (capacity < 0 || nchar < 0)
+	{
+		Zero();
+		return;
```
Comment (jblomer) [inline core/base/src/TString.cxx:249]: `Init` needs to return a char pointer.

File: core/base/src/TString.cxx [hunk 3]
```diff
@@ -361,6 +371,11 @@ TString& TString::operator=(const TSubString &substr)
 
 TString& TString::Append(char c, Ssiz_t rep)
 {
+	if (rep < 0)
+	{
+		Zero();
```
Comment (jblomer) [inline core/base/src/TString.cxx:376]: In this case, we should not zero out the string but append nothing.

File: core/base/src/TString.cxx [hunk 4]
```diff
@@ -402,6 +417,11 @@ TString& TString::Append(char c, Ssiz_t rep)
 
 Ssiz_t TString::Capacity(Ssiz_t nc)
 {
+	if (nc < 0)
```
Comment (jblomer) [inline core/base/src/TString.cxx:420]: We don't need any special handling here.  A negative `nc` is in particular smaller than `Length()`, therefore resulting in a no-op.

File: core/base/test/TString.cxx [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+#include "gmock/gmock.h"
+
+#include "TString.h"
+
+TEST(TString)
+{
+	TString n("testNeg", -1);
+	TString p("testPos", 1);
+	EXPECT_STREQ("", n);
+	EXPECT_STREQ("te", p);
```
Comment (jblomer) [inline core/base/test/TString.cxx:11]: Wouldn't we expect it to be `t`?

**Review by jblomer — CHANGES_REQUESTED** [review]
Please see the comments inline.

For some general code style issues (e.g. spacing), please see #2298 

The unit test would need to be referenced in the CMakeLists.txt file in the same directory in order to be built.

File: core/base/test/TString.cxx (hunk 5 above)
Comment (AdvaitDhingra) [reply core/base/test/TString.cxx:11]: Oh, I thought the first index was 0. I'll change it.

**Comment by AdvaitDhingra** [comment]
Thank you for all the help and patience guys. I'm learning a lot.

**Comment by Axel-Naumann** [comment]
Please recreate this PR with only your commits, by creating a new branch off master and then opening a new PR with that branch. This PR here has unrelated commits in it. Please also keep in mind that we do not use merge commits; please rebase your feature branch onto master instead.

**Pull request closed without merging**